File: frontends/kde/src/branding.cpp
#include "branding.h"

#include <QByteArray>
#include <QFile>
#include <QJsonDocument>
#include <QRegularExpression>

namespace branding {

namespace {

const char *kEnvFile = "BOOTC_INSTALLER_BRANDING";
const char *kEnvName = "BOOTC_INSTALLER_PRODUCT_NAME";
const char *kNeutralName = "Linux";
const char *kNeutralId = "linux";

const QStringList &brandingPaths()
{
    static const QStringList paths = {
        QStringLiteral("/run/host/etc/bootc-installer/branding.json"),
        QStringLiteral("/run/host/usr/share/bootc-installer/branding.json"),
        QStringLiteral("/etc/bootc-installer/branding.json"),
        QStringLiteral("/usr/share/bootc-installer/branding.json"),
    };
    return paths;
}

const QStringList &osReleasePaths()
{
    static const QStringList paths = {
        QStringLiteral("/run/host/etc/os-release"),
        QStringLiteral("/run/host/usr/lib/os-release"),
        QStringLiteral("/etc/os-release"),
        QStringLiteral("/usr/lib/os-release"),
    };
    return paths;
}

// The subset of shell quoting os-release allows: double quotes with backslash
// escapes, single quotes, or a bare word that ends at the first whitespace.
QString unquoteShell(const QString &value)
{
    if (value.isEmpty())
        return QString();
    const QChar first = value.at(0);
    if (first == QLatin1Char('"')) {
        QString out;
        bool escaped = false;
        for (int i = 1; i < value.size(); ++i) {
            const QChar c = value.at(i);
            if (escaped) {
                out.append(c);
                escaped = false;
            } else if (c == QLatin1Char('\\')) {
                escaped = true;
            } else if (c == QLatin1Char('"')) {
                break;
            } else {
                out.append(c);
            }
        }
        return out.trimmed();
    }
    if (first == QLatin1Char('\'')) {
        const QString rest = value.mid(1);
        const int end = rest.indexOf(QLatin1Char('\''));
        return (end >= 0 ? rest.left(end) : rest).trimmed();
    }
    const int ws = value.indexOf(QRegularExpression(QStringLiteral("\\s")));
    return ws >= 0 ? value.left(ws) : value;
}

QString pick(const QJsonObject &file, const QHash<QString, QString> &osr, const QString &key,
             const QStringList &osKeys, const QString &def)
{
    const QString fromFile = file.value(key).toString().trimmed();
    if (!fromFile.isEmpty())
        return fromFile;
    for (const QString &k : osKeys) {
        const QString v = osr.value(k).trimmed();
        if (!v.isEmpty())
            return v;
    }
    return def;
}

} // namespace

QHash<QString, QString> parseOsRelease(const QString &text)
{
    QHash<QString, QString> out;
    const QStringList lines = text.split(QLatin1Char('\n'));
    for (const QString &raw : lines) {
        const QString line = raw.trimmed();
        if (line.isEmpty() || line.startsWith(QLatin1Char('#')))
            continue;
        const int eq = line.indexOf(QLatin1Char('='));
        if (eq < 0)
            continue;
        const QString key = line.left(eq).trimmed();
        if (key.isEmpty())
            continue;
        out.insert(key, unquoteShell(line.mid(eq + 1).trimmed()));
    }
    return out;
}

Branding fromSources(const QJsonObject &file, const QString &osReleaseText, const QString &nameOverride)
{
    const QHash<QString, QString> osr = parseOsRelease(osReleaseText);
    Branding b;
    b.name = pick(file, osr, QStringLiteral("name"),
                  {QStringLiteral("PRETTY_NAME"), QStringLiteral("NAME")}, QLatin1String(kNeutralName));
    b.id = pick(file, osr, QStringLiteral("id"), {QStringLiteral("ID")}, QLatin1String(kNeutralId));
    b.vendor = pick(file, osr, QStringLiteral("vendor"),
                    {QStringLiteral("VENDOR_NAME"), QStringLiteral("NAME")}, QString());
    b.homeUrl = pick(file, osr, QStringLiteral("home_url"), {QStringLiteral("HOME_URL")}, QString());
    b.docsUrl = pick(file, osr, QStringLiteral("docs_url"), {QStringLiteral("DOCUMENTATION_URL")}, QString());
    b.supportUrl = pick(file, osr, QStringLiteral("support_url"), {QStringLiteral("SUPPORT_URL")}, QString());
    b.logo = pick(file, osr, QStringLiteral("logo"), {QStringLiteral("LOGO")}, QString());
    b.defaultHostname = pick(file, osr, QStringLiteral("default_hostname"),
                             {QStringLiteral("DEFAULT_HOSTNAME"), QStringLiteral("ID")}, QLatin1String(kNeutralId));
    b.defaultImage = pick(file, osr, QStringLiteral("default_image"), {}, QString());
    const QString overrideName = nameOverride.trimmed();
    if (!overrideName.isEmpty())
        b.name = overrideName;
    return b;
}

Branding resolveFrom(const QStringList &brandingPathList, const QStringList &osReleasePathList,
                     const QString &nameOverride)
{
    QJsonObject file;
    for (const QString &path : brandingPathList) {
        QFile f(path);
        if (!f.open(QIODevice::ReadOnly | QIODevice::Text))
            continue;
        const QJsonDocument doc = QJsonDocument::fromJson(f.readAll());
        if (!doc.isObject())
            continue;
        file = doc.object();
        break;
    }
    QString osRelease;
    for (const QString &path : osReleasePathList) {
        QFile f(path);
        if (!f.open(QIODevice::ReadOnly | QIODevice::Text))
            continue;
        const QString text = QString::fromUtf8(f.readAll());
        if (text.trimmed().isEmpty())
            continue;
        osRelease = text;
        break;
    }
    return fromSources(file, osRelease, nameOverride);
}

Branding resolve()
{
    QStringList paths = brandingPaths();
    const QString explicitPath = QString::fromUtf8(qgetenv(kEnvFile)).trimmed();
    if (!explicitPath.isEmpty())
        paths = {explicitPath};
    return resolveFrom(paths, osReleasePaths(), QString::fromUtf8(qgetenv(kEnvName)));
}

} // namespace branding

File: frontends/xfce/tuna_installer_xfce/branding.py
"""Product branding: a branding.json first, os-release second, neutral last.

Canonical copy: shared/branding/branding.py. bootc_installer/utils/branding.py
and frontends/xfce/tuna_installer_xfce/branding.py are byte-identical copies
(tests/unit/test_shared_branding.py enforces it). The contract, the key
table and the fixtures the other languages test against are in
shared/branding/README.md.

No GTK, no gettext, nothing but the standard library, so every frontend and
every test can import it as-is.
"""

import json
import os
import shlex
from dataclasses import asdict, dataclass

ENV_FILE = "BOOTC_INSTALLER_BRANDING"
ENV_NAME = "BOOTC_INSTALLER_PRODUCT_NAME"

# Host first: inside a Flatpak /etc is the runtime's, the host's is under
# /run/host. The first readable file wins; files are not merged.
BRANDING_PATHS = [
    "/run/host/etc/bootc-installer/branding.json",
    "/run/host/usr/share/bootc-installer/branding.json",
    "/etc/bootc-installer/branding.json",
    "/usr/share/bootc-installer/branding.json",
]
OS_RELEASE_PATHS = [
    "/run/host/etc/os-release",
    "/run/host/usr/lib/os-release",
    "/etc/os-release",
    "/usr/lib/os-release",
]

NEUTRAL_NAME = "Linux"
NEUTRAL_ID = "linux"

KEYS = (
    "name", "id", "vendor", "home_url", "docs_url", "support_url",
    "logo", "default_hostname", "default_image",
)


@dataclass(frozen=True)
class Branding:
    name: str = NEUTRAL_NAME
    id: str = NEUTRAL_ID
    vendor: str = ""
    home_url: str = ""
    docs_url: str = ""
    support_url: str = ""
    logo: str = ""
    default_hostname: str = NEUTRAL_ID
    default_image: str = ""
    # Where `name` came from: "env", "file", "os-release" or "default".
    # Diagnostics only; not part of the cross-language contract.
    source: str = "default"

    def as_dict(self) -> dict:
        d = asdict(self)
        d.pop("source")
        return d


def parse_os_release(text: str) -> dict:
    """KEY=value pairs from os-release text, quotes and escapes removed.

    Lines that do not parse are skipped rather than failing the whole file:
    a stray quote in a comment must not cost the product its name.
    """
    out = {}
    for raw in text.splitlines():
        line = raw.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        key, value = line.split("=", 1)
        key = key.strip()
        value = value.strip()
        try:
            parts = shlex.split(value)
            value = parts[0] if parts else ""
        except ValueError:
            value = value.strip("\"'")
        if key:
            out[key] = value.strip()
    return out


def _clean(value) -> str:
    return value.strip() if isinstance(value, str) else ""


def from_sources(file_data, os_release_text, name_override="") -> Branding:
    """Pure resolution: the merge rule from README.md, no filesystem access.

    file_data: the parsed branding.json (a dict) or None.
    os_release_text: os-release contents or None.
    """
    file_data = file_data if isinstance(file_data, dict) else {}
    osr = parse_os_release(os_release_text) if os_release_text else {}

    def pick(key, *os_keys, default=""):
        value = _clean(file_data.get(key))
        if value:
            return value, "file"
        for os_key in os_keys:
            value = _clean(osr.get(os_key))
            if value:
                return value, "os-release"
        return default, "default"

    name, source = pick("name", "PRETTY_NAME", "NAME", default=NEUTRAL_NAME)
    override = _clean(name_override)
    if override:
        name, source = override, "env"
    id_, _ = pick("id", "ID", default=NEUTRAL_ID)
    vendor, _ = pick("vendor", "VENDOR_NAME", "NAME")
    home_url, _ = pick("home_url", "HOME_URL")
    docs_url, _ = pick("docs_url", "DOCUMENTATION_URL")
    support_url, _ = pick("support_url", "SUPPORT_URL")
    logo, _ = pick("logo", "LOGO")
    default_hostname, _ = pick("default_hostname", "DEFAULT_HOSTNAME", "ID", default=NEUTRAL_ID)
    default_image, _ = pick("default_image")
    return Branding(
        name=name, id=id_, vendor=vendor, home_url=home_url, docs_url=docs_url,
        support_url=support_url, logo=logo, default_hostname=default_hostname,
        default_image=default_image, source=source,
    )


def _read_text(path: str):
    try:
        with open(path, encoding="utf-8", errors="replace") as fh:
            return fh.read()
    except OSError:
        return None


def _read_branding_file(paths) -> dict | None:
    for path in paths:
        text = _read_text(path)
        if text is None:
            continue
        try:
            data = json.loads(text)
        except ValueError:
            continue
        if isinstance(data, dict):
            return data
    return None


def resolve(branding_paths=None, os_release_paths=None, env=None) -> Branding:
    """The branding for this machine. Every argument exists for tests."""
    env = os.environ if env is None else env
    if branding_paths is None:
        explicit = _clean(env.get(ENV_FILE))
        branding_paths = [explicit] if explicit else BRANDING_PATHS
    if os_release_paths is None:
        os_release_paths = OS_RELEASE_PATHS

    file_data = _read_branding_file(branding_paths)
    os_release_text = None
    for path in os_release_paths:
        text = _read_text(path)
        if text is not None and text.strip():
            os_release_text = text
            break
    return from_sources(file_data, os_release_text, env.get(ENV_NAME, ""))
